ocaml/xapi/xapi_psr.ml

```diff
@@ -186,24 +187,123 @@ functor
 
 let perm = 0o640
 
-let cleanup_internal ~additional_files_to_remove () =
-  if List.length !Xapi_globs.pool_secrets <= 1 then
-    D.info "xapi_psr.ml:cleanup_internal: already cleaned up"
-  else
-    (* if we don't remove pool secret backups, they will be loaded into memory when xapi starts *)
-    let files_to_remove =
-      List.append additional_files_to_remove
-        [old_pool_secret_backup_path; new_pool_secret_backup_path]
+(* we include some sanity checks for the following invariants:
+     - only the master should have a checkpoint
+     - either (a) all the master psr state exists (=> resuming a failed psr)
+       or     (b) none of the state exists (=> starting a new psr)
+     - existing pool secret backups are consistent with requested pool secret
+       changes
+     - the runtime state, i.e. Xapi_globs.pool_secrets (although this check
+       is not done here, but at each stage separately)
+   these invariants could be broken if for example a psr fails, the master
+   changes, and then a new psr starts on the new master
```

Comment:
You should make pool secret rotation and HA be mutually exclusive.
Master promotion can occur at any time when HA is enabled (it cannot be blocked, it is essential to the functioning of HA), so there should be an assert that HA is not enabled, and raise `Api_errors.ha_is_enabled,` otherwise if one attempts to initiate a pool secret rotation. I see this is already present in `xapi_psr.ml`.

Similarly in the HA enable function there should be a check that no pool secret rotation is in progress.
I think this'll need a new operation in allowed operations for the pool in xapi_pool_helpers, and to remove `ha_enable` when psr becomes active.

Reply:
I opted for adding a flag to the database to solve this - that way we can block HA in between a failed PSR and the next attempt. I didn't see much benefit adding PSR to the operations list as well as this flag